keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,510 @@
+Non IP routable traffic, past the proxy, would need to be a non Kubernetes mechanism to route.
+
+Running the connectivity proxy in a separate process has a few advantages.
+- The connectivity proxy can be extended without recompiling the KAS.
+Administrators can run their own variants of the connectivity proxy.
+- Traffic can be audited or forwarded (eg. via a proprietary VPN) using a custom connectivity proxy.
+- The separation removes master <-> cluster connectivity concerns from the KAS.
+- The code and responsibility separation lowers the complexity of the KAS code base.
+- The separation reduces the effects of issue such as crashes in the connectivity impacting the KAS.
+Connectivity issues will not stop the KAS from serving API requests.
+This is important as serving API requests may be necessary in order to fix the crashes.
+A problem with a node, set of nodes or load-balancers configuration, may be fixed with API requests.
+
+![API Server Network Proxy Simple Cluster](NetworkProxySimpleCluster.png)
+The diagram shows API Server’s outgoing traffic flow.
+The user (in blue box), master network (in purple cloud) and
+a cluster network (in green cloud) are represented.
+
+The user (blue) initiates communication to the KAS.
+The KAS then initiates connections to other components.
+It could be node/pod/service in cluster networks (red dotted arrow to green cloud),
+or etcd for storage in the same master network (blue arrow) or mutate the request
+based on an admission web-hook (red dotted arrow to purple cloud).
+The KAS handles these cases based on NetworkContext based traffic routing.
+The connectivity proxy should be able to do routing solely based on IP.
+The proxy should not require the NetworkContext. This means the service CIDR,
+node CIDR and pod CIDR of each cluster network cannot overlap.
+
+### Network Context
+
+The minimal NetworkContext looks like the following struct in golang:
+
+```go
+type NetworkContext struct {
+        // NetworkName is the unique name of the network
+        // to select to route the traffic to.
+        NetworkName string
+}
+```
+
+NetworkName specifies the network to route traffic to.
+The KAS starts with a list of registered network names and
+the corresponding Connectivity services. So the KAS knows where to
+proxy the traffic to, otherwise it return an “Unknown network” error.
+
+The KAS starts with a proxy configuration like the below example.
+The example specifies 4 networks. "direct" specifies the KAS talking directly
+on the local network (no proxy). "master" specifies the KAS talks to a proxy
+listening at 1.2.3.4:5678. "cluster" specifies the KAS talk to a proxy
+listening at 1.2.3.5:5679. While these are represented as resources
+they are not intended to be loaded dynamically.
+The KAS loads them as configuration at start time.
+
+```yaml
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
```

Comment:
Shouldn't the name of the ConnectivityServiceConfiguration be the name referenced above in `networkName`?

that is we would have:

```yaml
apiVersion: ...
kind: ...
name: direction
spec:
  ...
```

Would help if it is a resource to match k8s resource format, or if it is not so say so explicitly? Also, helps to have a golang struct sketch if possible.

Reply:
Fixed